Core/Controllers/Api/ReportApiController.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using System.Linq;
+
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Http;
+
+using VoteMyst.Database;
+using VoteMyst.Controllers.Api.Models;
+
+namespace VoteMyst.Controllers.Api
+{
+    public class ReportApiController : VoteMystApiController
+    {
+        public ReportApiController(IServiceProvider serviceProvider) : base(serviceProvider) { }
+
+        [HttpPost]
+        [Route("api/reports/submit")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        public IActionResult SubmitReport([FromBody] ReportSubmission reportSubmission)
+        {
+            if (!User.Identity.IsAuthenticated)
+                return Unauthorized();
+
+            Entry targetEntry = DatabaseHelpers.Context.QueryByDisplayID<Entry>(reportSubmission.EntryDisplayId);
+            if (targetEntry == null)
+                return BadRequest("The specified entry could not be found.");
+
+            UserAccount currentUser = GetCurrentUser();
+
+            if (currentUser.ID == targetEntry.Author.ID)
+                return BadRequest("Reporting own posts is not allowed.");
+            if (targetEntry.Reports.Any(r => r.ReportAuthor.ID == currentUser.ID))
+                return BadRequest("A report on this entry by the current user already exists.");
+
+            DatabaseHelpers.Entries.ReportEntry(targetEntry, currentUser, reportSubmission.Reason);
+
+            return Ok();
+        }
+
+        [HttpPost]
+        [Route("api/reports/{id}/approve")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        public IActionResult DeleteReportedPost(int id)
+            => ProcessReportAction(id, ReportStatus.Approved);
+
+        [HttpPost]
+        [Route("api/reports/{id}/reject")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        public IActionResult RejectReport(int id)
+            => ProcessReportAction(id, ReportStatus.Rejected);
+
+        private IActionResult ProcessReportAction(int reportId, ReportStatus desiredStatus)
+        {
+            if (desiredStatus == ReportStatus.Pending)
+                throw new ArgumentException("The desired status 'Pending' cannot be processed.");
```

Comment:
What does this return to the user? 503 with the stacktrace?

Reply:
In theory this should never be reachable, since the call is only made by a JS script on our site, so the call can't really be made by the user. But I would assume a BadRequestResult?

Reply:
If someone tries to manually send the data can this block be reached? 

If it returns the stacktrace this should be changed. (Or we should be able to disable the stacktrace)

Reply:
Since this is not a view (has no graphical returns), it should only return the status code, along with the error message inside the body. Stacktraces are not included in that IIRC.